src/states/Preroll.js

```diff
@@ -0,0 +1,221 @@
+import videojs from 'video.js';
+
+import {AdState, ContentPlayback, BeforePreroll} from '../states.js';
+import cancelContentPlay from '../cancelContentPlay.js';
+import {startAdBreak, endAdBreak} from '../adBreak.js';
+
+/*
+ * This state encapsulates waiting for prerolls, preroll playback, and
+ * content restoration after a preroll.
+ */
+export default class Preroll extends AdState {
+
+  init(player, adsReady) {
+    // Loading spinner from now until ad start or end of ad break.
+    player.addClass('vjs-ad-loading');
+
+    // Start the clock ticking for ad timeout
+    this._timeout = player.setTimeout(function() {
+      player.trigger('adtimeout');
+    }, player.ads.settings.prerollTimeout);
+
+    // If adsready already happened, lets get started. Otherwise,
+    // wait until onAdsReady.
+    if (adsReady) {
+      this.handleAdsReady();
+    } else {
+      this.adsReady = false;
+    }
+  }
+
+  onAdsReady(player) {
+    if (!player.ads.inAdBreak() && !player.ads.isContentResuming()) {
+      player.ads.debug('Received adsready event (Preroll)');
+      this.handleAdsReady();
+    } else {
+      videojs.log.warn('Unexpected adsready event (Preroll)');
+    }
+  }
+
+  /*
+   * Ad integration is ready. Let's get started on this preroll.
+   */
+  handleAdsReady() {
+    this.adsReady = true;
+    if (this.player.ads.nopreroll_) {
+      this.noPreroll();
+    } else {
+      this.readyForPreroll();
+    }
+  }
+
+  /*
+   * If there is no preroll, play content instead.
+   */
+  noPreroll() {
+    this.player.ads.debug('Skipping prerolls due to nopreroll event');
+    this.transitionTo(ContentPlayback);
+  }
+
+  /*
+   * Fire the readyforpreroll event. If loadstart hasn't happened yet,
+   * wait until loadstart first.
+   */
+  readyForPreroll() {
+    const player = this.player;
+
+    // Signal to ad plugin that it's their opportunity to play a preroll.
+    if (player.ads._hasThereBeenALoadStartDuringPlayerLife) {
+      player.ads.debug('Triggered readyforpreroll event (Preroll)');
+      player.trigger('readyforpreroll');
+
+    // Don't play preroll before loadstart, otherwise the content loadstart event
+    // will get misconstrued as an ad loadstart. This is only a concern for the
+    // initial source; for source changes the whole ad process is kicked off by
+    // loadstart so it has to have happened already.
+    } else {
+      player.one('loadstart', () => {
+        player.ads.debug('Triggered readyforpreroll event (loadstart)');
+        player.trigger('readyforpreroll');
+      });
+    }
+  }
+
+  /*
+   * Don't allow the content to start playing while we're dealing with ads.
+   */
+  onPlay(player) {
+    player.ads.debug('Received play event (Preroll)');
+
+    if (!this.inAdBreak() && !this.isContentResuming()) {
```

Comment:
should this be `!isInAdMode()`?

Reply:
isAdBreak is the new isInAdMode. Are you saying you don't think the content resuming check should be there? That has historically been what we've done, and I'm not aware of a reason to change it: I don't think you should pause the player during content resuming.

Reply:
no, I think the `!isContentResuming()` is likely to be correct. I think the `!inAdBreak()` part is a bit confusing since plays will not be cancelled while waiting for `startLinearAdMode()` while in the Preroll state, right?

Reply:
uh, sorry I meant plays will not be cancelled while the ad is actually playing

Reply:
The historical behavior has been that plays are not cancelled during an ad break. I believe this works out because the player controls are hidden, so the user cannot request a content play. A programatic call to play during ads would break things. Maybe we can improve this using middleware!